### PR TITLE
implement trainingjob controller without autoscaler

### PR DESCRIPTION
**Pull request by m3ngyang**
This pull request implements a training job controller without autoscaler, which registers a self-defined resource called `TrainingJob` and then watches the event of `TrainingJob` to manage the lifecycle of `TrainingJob` instance.

### CODE REVIEW

File: pkg/controller/trainingjob_controller.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+	paddlev1 "github.com/paddlepaddle/edl/pkg/apis/paddlepaddle/v1"
+	paddleclient "github.com/paddlepaddle/edl/pkg/client/clientset/versioned"
+	paddlescheme "github.com/paddlepaddle/edl/pkg/client/clientset/versioned/scheme"
+	paddleinformers "github.com/paddlepaddle/edl/pkg/client/informers/externalversions"
+	paddlelisters "github.com/paddlepaddle/edl/pkg/client/listers/paddlepaddle/v1"
+	"github.com/paddlepaddle/edl/pkg/updater"
+)
+
+type TrainingJobController struct {
```
Comment (typhoonzero) [inline pkg/controller/trainingjob_controller.go]: Exported type must have comments. Curious this should be checked by the linter.

File: pkg/controller/trainingjob_controller.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
+package controller
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
```
Comment (typhoonzero) [inline pkg/controller/trainingjob_controller.go]: We prefer to use [log15](https://github.com/inconshreveable/log15) recently.

File: pkg/controller/trainingjob_controller.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+	paddleInformers paddleinformers.SharedInformerFactory) *TrainingJobController {
+	trainingjobInformer := paddleInformers.Paddlepaddle().V1().TrainingJobs()
+	paddlescheme.AddToScheme(scheme.Scheme)
+
+	glog.V(4).Info("Creating trainingjob event broadcaster")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeCli.CoreV1().Events("")})
+	workqueue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "TrainingJob")
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "TrainingJobController"})
```
Comment (typhoonzero) [inline pkg/controller/trainingjob_controller.go]: Maybe we need some document or a sample controller to inform these concepts in a controller:

- informer
- event broadcaster
- workqueue
- recorder

File: pkg/controller/trainingjob_controller.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+// informer caches and starting workers. It will block until stopCh
+// is closed, at which point it will shutdown the workqueue and wait for
+// workers to finish processing their current work items.
+func (c *TrainingJobController) Run(threadiness int, stopCh <-chan struct{}) error {
+	// TODO add a lock to ensure there is only one controller in the cluster
+	defer runtime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	glog.Info("Starting trainingjob controller")
+	glog.Info("Starting to create custom resource definition")
```
Comment (typhoonzero) [inline pkg/controller/trainingjob_controller.go]: Maybe can reduce some info logs, output info log only when the operation success or fail.

File: pkg/controller/trainingjob_controller.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+	oldtj.Delete()
+	c.jobupdater.Delete(key)
+}
+
+func (c *TrainingJobController) runWorker() {
+	for c.processNestWorkItem() {
+	}
+}
+
+func (c *TrainingJobController) processNestWorkItem() bool {
```
Comment (typhoonzero) [inline pkg/controller/trainingjob_controller.go]: Looks like the name should be `processNextWorkItem`

Comment (m3ngyang) [reply pkg/controller/trainingjob_controller.go]: Yes, it's a typo here.

File: pkg/controller/trainingjob_controller.go (hunk 1 above)
Comment (m3ngyang) [reply pkg/controller/trainingjob_controller.go]: Right, I will supplement some comments.

File: pkg/controller/trainingjob_controller.go (hunk 2 above)
Comment (m3ngyang) [reply pkg/controller/trainingjob_controller.go]: Is there any difference between those two log packages?

File: pkg/controller/trainingjob_controller.go (hunk 4 above)
Comment (m3ngyang) [reply pkg/controller/trainingjob_controller.go]: Maybe we can use different log levels for them, abundant logs are useful when debugging.

File: pkg/signals/signal.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+	"os/signal"
+)
+
+var (
+	onlyOneSignalHandler = make(chan struct{})
+	shutdownSignals      = []os.Signal{os.Interrupt}
+)
+
+// SetupSignalHandler registered for SIGTERM and SIGINT. A stop channel is returned
+// which is closed on one of these signals. If a second signal is caught, the program
```
Comment (helinwang) [inline pkg/signals/signal.go:14]: "If a second signal is caught, the program is terminated with exit code 1." --- this is a hardcoded rule that people don't expect in general. Do we really need it? (if the user really want to terminate it, he can send SIGKILL using command `kill -9`).

For graceful shutdown, I think typically it's done in the following way:

1. setup signal handler
1. start a goroutine that does the real work, block main function
1. when signal is caught, trigger graceful shutdown (e.g., closing a channel, cancel a context).
1. the main function will return normally when graceful shutdown is complete.

An example: https://gist.github.com/peterhellberg/38117e546c217960747aacf689af3dc2

File: pkg/controller/trainingjob_controller.go [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+	return controller
+}
+
+// Run will set up the event handlers for trainingjob, as well as syncing
+// informer caches and starting workers. It will block until stopCh
+// is closed, at which point it will shutdown the workqueue and wait for
+// workers to finish processing their current work items.
+func (c *TrainingJobController) Run(threadiness int, stopCh <-chan struct{}) error {
+	// TODO add a lock to ensure there is only one controller in the cluster
```
Comment (helinwang) [inline pkg/controller/trainingjob_controller.go:108]: Please put your name after TODO, e.g., `// TODO(helin): ...`

Comment (helinwang) [inline pkg/controller/trainingjob_controller.go:108]: Do we need this TODO? Maybe it's the job of the binary (e.g., `cmd/paddle_controller/paddle_controller.go`) to ensure only one instance is running, rather than the library.

File: pkg/controller/trainingjob_controller.go [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+	}
+
+	log.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.trainingjobSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	log.Info("Starting workers")
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
```
Comment (helinwang) [inline pkg/controller/trainingjob_controller.go:125]: Why do we need a worker pool? Goroutines are already multiplexed onto a thread pool by the Go runtime, perhaps we can have a single loop, and start a new goroutine for each iteration.

File: pkg/controller/trainingjob_controller.go (hunk 7 above)
Comment (m3ngyang) [reply pkg/controller/trainingjob_controller.go:108]: This comment is redundant, and I'll delete it. A leader election has been implemented to assure that there is only one controller to manage TrainingJob.

File: pkg/controller/trainingjob_controller.go (hunk 8 above)
Comment (m3ngyang) [reply pkg/controller/trainingjob_controller.go:125]: Sorry, I don't understand the differences between the worker pool and the single loop. Here, `wait.Util(f func(), period time.Duration, stopCh <-chan struct{})` loops until stop channel is closed, running f every period.

Comment (helinwang) [reply pkg/controller/trainingjob_controller.go:125]: No worries, let me explain.

Here is the current code:

```go
for i := 0; i < threadiness; i++ {
	go wait.Until(c.runWorker, time.Second, stopCh)
}

func (c *TrainingJobController) runWorker() {
	for c.processNextWorkItem() {
	}
}
```

And there are *threadiness* number of gorountines running all work items (assume *N*). The max concurrency is *threadiness*.

Another implementation is:

```go
go wait.Until(c.runWorker, time.Second, stopCh) // only a single wait.Until

func (c *TrainingJobController) runWorker() {
	for item := range c.itemCh {
                go c.process(item)
	}
}
```

In this way the The max concurrency is *N* rather than *threadiness*. Eliminating the unnecessary configuration value *threadiness*, at the same time provides better concurrency.

In C++ developers usually uses a thread pool, but in Go, the runtime already use a thread pool to very efficiently run Goroutines, based on a very good scheduling algorithm. We don't have to worry about it.

Does it makes sense to you?

Comment (m3ngyang) [reply pkg/controller/trainingjob_controller.go:125]: Thanks for your elaborated explanation, and I consulted implementation of some built-in controllers, such as [deployment controller](https://github.com/kubernetes/kubernetes/blob/97fdf8ac70dfdaf07074dee6b6cb345be71e96ec/pkg/controller/deployment/deployment_controller.go#L148) and [job controller](https://github.com/kubernetes/kubernetes/blob/97fdf8ac70dfdaf07074dee6b6cb345be71e96ec/pkg/controller/job/job_controller.go#L137). These controllers usually use a controlled number to indicate the max number of objects that are allowed to sync concurrently, instead of the total number. Besides, sync number can be managed by the [configuration of controller-manager](https://kubernetes.io/docs/reference/generated/kube-controller-manager/), for example `--concurrent-deployment-syncs int32`.
To sum up, I think it's more reasonable to use *threadiness* as the max concurrency.

Comment (helinwang) [reply pkg/controller/trainingjob_controller.go:125]: I see, I did not realize that there is a need for limiting the concurrency, now it makes sense. Thanks!

**Comment by m3ngyang** [comment]
this pr is included in #24 
closed here.
